Make the plugin work

SamCarlberg opened the pull request.

SamCarlberg committed:
Make plugin work

SamCarlberg committed:
Remove default jlink tasks, add functions for generating task names

SamCarlberg committed:
Do not use afterEvaluate when generating jlink tasks

SamCarlberg committed:
Add check for module linking

SamCarlberg committed:
Use properties for JLinkTask fields

Breaks task caching

SamCarlberg committed:
Use providers when binding jlink tasks to options objects

Fixes task caching, but at what cost?

SamCarlberg committed:
Move options copying to task registration

No more eager initialization, no sir

JLLeitschuh committed:
Make sample check task depend upon included build check task

SamCarlberg committed:
Fix jlink zip, run jdeps when required

SamCarlberg committed:
Add docs to JLinkOptions.applicationJar

SamCarlberg committed:
Make sample-app a trivial application with jlink support

Requires manually setting jlink task dependencies, but that will be fixed later

SamCarlberg committed:
Scan project and its dependencies to determine modules to link

Modules are determined by analyzing the libraries the project depends on, as well as running jdeps on build/classes directory to determine the modules used directly by the project
Remove the modules and modulePath options, since they are no longer needed
Add a sample JavaFX app project

SamCarlberg committed:
Generate .zip archives on windows, .tar.gz on unix

JLLeitschuh committed:
Refactor plugin registration to be lazily evaluated

JLLeitschuh committed:
Simplify the unit test logic by using the plugins block instead

SamCarlberg committed:
Merge pull request #3 from GradleWeaver/tests-improvements

Simplify the unit test logic by using the plugins block instead

JLLeitschuh committed:
Add Travis CI configuration to build

SamCarlberg committed:
Fix tests, make root check task rely on sample project jlink execution

JLLeitschuh committed:
Don't test with jdk 8, do test with jdk 9

SamCarlberg committed:
Remove use of Java 10 API to allow building on Java 9

SamCarlberg committed:
Add readme

SamCarlberg committed:
Run Travis with stacktrace enabled

SamCarlberg committed:
Update readme with further information on available jlink options

SamCarlberg committed:
Allow users to specify extra modules to link 

Useful if the application or libraries it uses reference some class reflectively, since jdeps can't detect reflective references to modules

SamCarlberg committed:
Lazily evaluate Java bin directory

SamCarlberg committed:
Move enums to their own files

SamCarlberg committed:
Update readme to mention JDK 9 is not supported

Also mention the extraModules option

SamCarlberg committed:
Remove OpenJDK 9 and OracleJDK 10 from Travis config

OpenJDK 10 and 11 are still running

SamCarlberg committed:
Add launcher scripts

SamCarlberg committed:
Move launcher script generation to JlinkTask

Launcher scripts will be generated if launcher options are set. No need for a separate task

SamCarlberg committed:
Give jlink configurations their own build directories

So building images from multiple configurations don't overwrite eachother

SamCarlberg committed:
Update readme with launcher DSL

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkExtension.kt:
```diff
@@ -9,7 +10,39 @@ open class JLinkExtension
         project: Project
 ) {
-    val configure: NamedDomainObjectCollection<JLinkOptions> = project.container(JLinkOptions::class.java) { name ->
+    val configure: NamedDomainObjectContainer<JLinkOptions> = project.container(JLinkOptions::class.java) { name ->
```
This needs to be renamed to avoid the name collision.
https://github.com/gradle/kotlin-dsl/issues/1175

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkExtension.kt:
````diff
@@ -9,7 +10,39 @@ open class JLinkExtension
+     * }
+     * ```
+     */
+    operator fun String.invoke(configurationAction: Action<in JLinkOptions>) = register(this, configurationAction)
````
Do we really want to have this method in our public API. It won't be usable from a Groovy DSL build script.

SamCarlberg replied:
But it _is_ usable from a kotlin build script, and makes the kotlin configuration easier IMO:

```kotlin
jlink {
  "configuration name" {
    // configuration settings
  }
}
```

versus

```kotlin
jlink {
  register("configuration name") {
    // configuration settings
  }
}
```

JLLeitschuh replied:
We should be exposing the `NamedDomainObjectContainer`'s API as the primary way for adding `JLinkOptions`. The `NamedDomainObjectContainer` is a standard API so it's better to just let users use that without providing helper methods for interacting with it. IMO.

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkExtension.kt:
```diff
@@ -9,7 +10,39 @@ open class JLinkExtension
+     * @param configurationAction the action to use to set up the configuration
+     */
+    fun register(name: String, configurationAction: Action<in JLinkOptions>)
+            = configure.register(name, configurationAction)
```
Same question here.

SamCarlberg replied:
I don't see how this isn't usable from a Groovy DSL

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkLauncherOptions.kt:
```diff
@@ -0,0 +1,48 @@
+
+import java.io.Serializable
+
+class JLinkLauncherOptions : Serializable {
```
Why `Serializable`?

SamCarlberg replied:
Gradle was throwing `NotSerializableException` when this was being used. It doesn't seem to be happening now, though... that's really weird. I guess I can remove this, then

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkLauncherOptions.kt:
```diff
@@ -0,0 +1,48 @@
+
+    fun setVmOptions(options: Iterable<String>) {
+        vmOptions = options.toMutableList()
+    }
```
Why have this extra method?

SamCarlberg replied:
So VM options can be more easily copied from the `application` plugin configuration:

```kotlin
application {
  applicationDefaultJvmArgs = listOf(...)
}

jlink {
  "configuration" {
    vmOptions(application.applicationDefaultJvmArgs)
  }
}
```

JLLeitschuh replied:
Isn't that exactly the same as `vmOptions = application.applicationDefaultJvmArgs`?
Also, if you want that, this should really be:
```kotlin
vmOptions { application.applicationDefaultJvmArgs }
```
It would be better to make the `vmOptions` a `ListProperty`. That way the user can do:
```kotlin
vmOptions.set(provider { application.applicationDefaultJvmArgs} )
```

SamCarlberg replied:
That precludes users from adding more than just the JVM args from the application plugin, eg

```kotlin
jlink {
  "configuration" {
    vmOptions(application.applicationDefaultJvmArgs)
    vmOptions("-Xms16m", "-Xmx2G")
  }
}
```

JLLeitschuh replied:
Not if using a `ListProperty`. `ListProperty` has methods like `addAll`.

SamCarlberg committed:
Remove serializable from launcher options class

SamCarlberg committed:
Make linux launch scripts work for configurations with spaces

SamCarlberg committed:
Add a sample with the Groovy DSL

JLLeitschuh commented on README.md:
```diff
@@ -0,0 +1,59 @@
+
+jlink {
+  "configuration name" {
+    applicationJar = tasks.getByName<Jar>("jar").archivePath // No default value
```
This forces the configuration of the "jar" task, which is not good.
Additionally, if a configuration block is executed later that further configures all the `Jar` tasks, this won't update. We shouldn't do this.

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkPlugin.kt:
```diff
@@ -10,21 +14,72 @@ open class JLinkPlugin : Plugin<Project> {
+                compression = Compression.GZIP
+                extension = "tar.gz"
+            }
+        }
```
Why generate a different artifact per OS? Why not just generate both? Also, why default to GZIP? The user can configure that easily if they want to.

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkOptions.kt:
```diff
@@ -1,12 +1,22 @@
+     * The application JAR. This will be copied into the generated jlink image's `bin` directory, next to the
+     * `java` executable.
+     */
+    var applicationJar: File? = null
```
This should be a `RegularFileProperty`.

SamCarlberg replied:
`RegularFileProperty` or just `Provider<File>`? `RegularFileProperty` is difficult to bind to normal `File` objects like those provided by the outputs of the `Jar` task

JLLeitschuh replied:
You want `RegularFileProperty`. I've fixed the `Jar` task for Gradle 5.1 to support being able to produce a `RegularFileProperty` directly.

SamCarlberg replied:
I'd prefer not to have to depend on bleeding edge releases. It limits the usability of the plugin

JLLeitschuh replied:
Regardless, we should still just use the `RegularFileProperty`.

SamCarlberg replied:
`RegularFileProperty` is impossible to use in 4.10 when setting it to a normal `File `object like `Jar::archivePath` or `ShadowJar::archivePath`. And as I said before, I don't want to depend on a bleeding edge release and limit the usability of the plugin

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkPlugin.kt:
```diff
@@ -10,21 +14,72 @@ open class JLinkPlugin : Plugin<Project> {
         const val JLINK_TASK_GROUP = "JLink"
         const val JLINK_TASK_NAME = "jlinkGenerate"
-        const val JLINK_ZIP_TASK_NAME = "jlinkZip"
+        const val JLINK_ARCHIVE_TASK_NAME = "jlinkArchive"
```
If you want these to be statically accessible from the Groovy DSL or from a different java based plugin, these need to be annotated with `@JvmStatic`.

SamCarlberg replied:
That's not required for `const`-qualified fields according to IntelliJ:
> '@JvmStatic' annotation is useless for const or '@JvmField' properties

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkTask.kt:
```diff
@@ -4,132 +4,257 @@ import org.gradle.api.DefaultTask
-    @OutputDirectory
-    var jlinkDir: DirectoryProperty = newOutputDirectory()
+    @get:InputFile
+    val applicationJarLocation: RegularFileProperty = newInputFile()
```
Apparently, `newInputFile` is deprecated in Gradle 5.0.

SamCarlberg replied:
Welp. As long as it'll still work

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkTask.kt:
```diff
@@ -4,132 +4,257 @@ import org.gradle.api.DefaultTask
-import org.gradle.api.tasks.OutputDirectory
-import org.gradle.api.tasks.TaskAction
+import org.gradle.api.tasks.*
 import org.gradle.internal.jvm.Jvm
```
❌  Internal API usage.

SamCarlberg replied:
What alternatives do you propose?

JLLeitschuh replied:
What's the use case or what do you need?

SamCarlberg replied:
To get the build JDK to determine the location of the `jlink` and `jdeps` tools. The `java.home` system property can't be used since Gradle may be running on a different JDK, like if `-Dorg.gradle.java.home` is set

JLLeitschuh replied:
Can you open an issue with Gradle or ask about an alternative solution in the Gradle Slack channel?

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkTask.kt:
```diff
@@ -4,132 +4,257 @@ import org.gradle.api.DefaultTask
-import org.gradle.api.tasks.TaskAction
+import org.gradle.api.tasks.*
 import org.gradle.internal.jvm.Jvm
+import org.gradle.internal.os.OperatingSystem
```
❌ Internal API usage.

SamCarlberg replied:
Unfortunately, this is a very useful API with no real public alternative. I could copy the class into the plugin, but that seems hacky

JLLeitschuh replied:
https://github.com/gradle/gradle/issues/7735

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/LauncherGenerator.kt:
```diff
@@ -0,0 +1,102 @@
+package org.gradleweaver.plugins.jlink
+
+import org.gradle.internal.os.OperatingSystem
```
❌ Internal API usage.

JLLeitschuh commented on plugin/src/test/kotlin/org/gradleweaver/plugins/jlink/LauncherGeneratorTest.kt:
```diff
@@ -0,0 +1,43 @@
+    @Test
+    fun testLinuxModule() {
+        val scriptText = generator.generateModuleLaunchScript(os = OperatingSystem.LINUX, moduleName = "mod", mainClassName = "main")
+        println(scriptText)
```
Shouldn't you assert something here?

SamCarlberg replied:
Whoops, forgot about that :sweat_smile:

JLLeitschuh commented:
Can we make it so that it's possible to build a windows Jlink executable from a Mac and vice versa?

SamCarlberg committed:
Use properties for jlink options

SamCarlberg committed:
Fix launcher generator test

SamCarlberg committed:
Rename "configure" to "jlinkConfigurations"

SamCarlberg committed:
Provide both .zip and .tar.gz archives on all build platforms

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/Endianness.kt:
```diff
@@ -0,0 +1,21 @@
+    /**
+     * Use the endianness of the build system.
+     */
+    SYSTEM_DEFAULT,
```
Shouldn't we just use the system default by default? Or is this because we can't determine the system default?

SamCarlberg replied:
This is equivalent to not passing the `--endian` flag. The `jlink` tool will use the byte order of the build system.

JLLeitschuh replied:
Okay.

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/CompressionLevel.kt:
```diff
@@ -0,0 +1,21 @@
+/**
+ * Options for the levels of compression available for generated runtime images.
+ */
+enum class CompressionLevel {
```
Can you link to documentation describing these each in more detail?

SamCarlberg replied:
https://docs.oracle.com/javase/9/tools/jlink.htm

SamCarlberg replied:
This is also linked in the README

JLLeitschuh replied:
Can you link to the exact section?
https://docs.oracle.com/javase/9/tools/jlink.htm#GUID-CECAC52B-CFEE-46CB-8166-F17A8E9280E9__GUID-DF03D0FC-9390-4C17-BEEA-ECC2EA9B94F2

SamCarlberg replied:
The page doesn't support linking to the individual sections

SamCarlberg committed:
Copy internal OperatingSystem class into plugin

Converts it to Kotlin

SamCarlberg committed:
Remove vmOptions functions, use val instead of var

SamCarlberg committed:
Update examples in README

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkTask.kt:
```diff
@@ -2,134 +2,259 @@ package org.gradleweaver.plugins.jlink
-        commandBuilder.add("--ignore-signing-information")
-    }
+        if (compressionLevel.isPresent) {
+            commandBuilder.add("--compress=${compressionLevel.get().ordinal}")
```
❌ Don't use ordinal, add a constant in the enum.

SamCarlberg committed:
Ordinals are bad

SamCarlberg committed:
Generate launcher script in main jlink task exec

JLLeitschuh commented on README.md:
```diff
@@ -0,0 +1,59 @@
+    endianess.set(JLinkTask.Endianness.SYSTEM_DEFAULT)
+    ignoreSigningInformation.set(false)
+    excludeHeaderFiles.set(false)
+    excludeManPages .set(false)
```
```suggestion
    excludeManPages.set(false)
```

JLLeitschuh commented on README.md:
```diff
@@ -0,0 +1,59 @@
+`--disable-plugin pluginname`  
+`--limit-modules`  
+`--save-opts`  
+`@filename`
```
Elaborate on what this is supposed to mean.

SamCarlberg replied:
The assumption here is that people using a jlink plugin will have at least familiarized themselves with the jlink options from the reference page. We shouldn't have to explain standard tool options

JLLeitschuh replied:
Erm... I don't think that's a valid assumption.

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkLauncherOptions.kt:
```diff
@@ -0,0 +1,43 @@
+
+    fun getApplicationModuleName(): String = applicationModuleName.get()
+
+    fun getMainClassName(): String = mainClassName.get()
```
Any of these `get***` methods could throw an exception if the value isn't present, just FYI.

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkOptions.kt:
```diff
@@ -1,95 +1,87 @@
 
     /**
      * Optimize `Class.forName` calls to constant class loads.
      */
```
Elaborate? I have no idea what this means from a practical sense.

SamCarlberg replied:
It bypasses class lookups

```java
// Your code
Class<?> fooClass = Class.forName("com.example.Foo");

// Is optimized to
Class<?> fooClass = com.example.Foo.class;
```

JLLeitschuh replied:
Fascinating! 👍 Can you add a comment about that?

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkOptions.kt:
```diff
@@ -1,95 +1,87 @@
      * Excludes man pages from the generated image.
      */
-    var excludeManPages = false
+    val excludeManPages = objectFactory.property(false)
```
How are man pages even generated??

SamCarlberg replied:
It's man pages for the Java executable(s) IIRC. Man pages are never generated by jlink

JLLeitschuh replied:
How do you declare man pages? Is there a location that you need to declare them to get them bundled? If so, then we should document how a user can get man pages added, ie. what convention they need to follow.

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkTask.kt:
```diff
@@ -2,134 +2,260 @@ package org.gradleweaver.plugins.jlink
+     */
+    @get:Input
+    @get:Optional
+    val launcherOptions = objectFactory.property<JLinkLauncherOptions>()
```
Why hold the nested element? Are all of the sub-elements of the `JLinkLauncherOptions` treated as inputs?

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkTask.kt:
```diff
@@ -2,134 +2,260 @@ package org.gradleweaver.plugins.jlink
-            rename {
-                "${project.name.toLowerCase()}.jar"
+        // Copy the application JAR into the jlink bin directory, if specified
+        if (applicationJarLocation.isPresent) {
```
Wouldn't it be an exception if this weren't present?

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkOptions.kt:
```diff
@@ -1,95 +1,87 @@
+     * `java` executable.
      */
-    var modules: List<String> = listOf()
+    val applicationJar = objectFactory.property<File>()
```
This should be a `RegularFileProperty`. Just having this be of type `Property<File>` won't automatically transfer the dependency information from the creating class.

JLLeitschuh commented on README.md:
```diff
@@ -0,0 +1,59 @@
+
+jlink {
+  "configuration name" {
+    applicationJar.set(provider { tasks.getByName<Jar>("jar").archivePath }) // No default value
```
You might be able to do this:
```kotlin
applicationJar.set(tasks.named<Jar>("jar").map { it.archivePath })
```

That API should exist in Gradle 5.0. It will also transfer the dependency information as well which is what we really want.

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkTask.kt:
```diff
@@ -2,134 +2,260 @@ package org.gradleweaver.plugins.jlink
 
-    internal val options: JLinkOptions
+    private val objectFactory = project.objects
+    private val launcherGenerator = LauncherGenerator()
```
Annotate as `@Internal` so it is ignored by the task `"up-to-date"` checking. Also, why is the `LauncherGenerator` an object when all it has are functions that don't operate on any member variables?

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkLauncherOptions.kt:
```diff
@@ -0,0 +1,43 @@
+
+import org.gradle.api.Project
+
+class JLinkLauncherOptions(project: Project) {
```
The constructor should be internal.

JLLeitschuh commented on README.md:
```diff
@@ -0,0 +1,59 @@
+    // jlink executable options. See the reference page for details
+    bindServices.set(false)
+    compressionLevel.set(JLinkTask.CompressionLevel.NONE)
+    endianess.set(JLinkTask.Endianness.SYSTEM_DEFAULT)
```
```suggestion
    endianess.set(Endianness.SYSTEM_DEFAULT)
```

JLLeitschuh replied:
Want to fix this before merging?

JLLeitschuh commented on README.md:
```diff
@@ -0,0 +1,59 @@
+    
+    // jlink executable options. See the reference page for details
+    bindServices.set(false)
+    compressionLevel.set(JLinkTask.CompressionLevel.NONE)
```
```suggestion
    compressionLevel.set(CompressionLevel.NONE)
```

JLLeitschuh replied:
Want to fix this before merging?

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkPlugin.kt:
```diff
@@ -10,21 +13,68 @@ open class JLinkPlugin : Plugin<Project> {
 
+            // Some workarounds to allow the options to have the JAR file and jlink dir specified as File objects
+            // instead of Gradle Property objects
+            applicationJarLocation.set(layout.file(provider { options.applicationJar.orNull?.relativeTo(projectDir) }))
```
Why relative to?
Also, you should be able to just set this directly.

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkPlugin.kt:
```diff
@@ -10,21 +13,68 @@ open class JLinkPlugin : Plugin<Project> {
+
+    private fun JLinkTask.copyFromOptions(options: JLinkOptions) {
+        with(project) {
+            imageName.set(provider { options.name })
```
```suggestion
            imageName.set(options.name)
```

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkPlugin.kt:
```diff
@@ -10,21 +13,68 @@ open class JLinkPlugin : Plugin<Project> {
+    private fun JLinkTask.copyFromOptions(options: JLinkOptions) {
+        with(project) {
+            imageName.set(provider { options.name })
+            bindServices.set(provider { options.bindServices.orNull })
```
```suggestion
            bindServices.set(options.bindServices)
```

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkPlugin.kt:
```diff
@@ -10,21 +13,68 @@ open class JLinkPlugin : Plugin<Project> {
+        with(project) {
+            imageName.set(provider { options.name })
+            bindServices.set(provider { options.bindServices.orNull })
+            compressionLevel.set(provider { options.compressionLevel.orNull })
```
```suggestion
            compressionLevel.set(options.compressionLevel)
```

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkPlugin.kt:
```diff
@@ -10,21 +13,68 @@ open class JLinkPlugin : Plugin<Project> {
+            imageName.set(provider { options.name })
+            bindServices.set(provider { options.bindServices.orNull })
+            compressionLevel.set(provider { options.compressionLevel.orNull })
+            endianness.set(provider { options.endianness.orNull })
```
```suggestion
            endianness.set(options.endianness)
```

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkPlugin.kt:
```diff
@@ -10,21 +13,68 @@ open class JLinkPlugin : Plugin<Project> {
+            bindServices.set(provider { options.bindServices.orNull })
+            compressionLevel.set(provider { options.compressionLevel.orNull })
+            endianness.set(provider { options.endianness.orNull })
+            ignoreSigningInformation.set(provider { options.ignoreSigningInformation.orNull })
```
```suggestion
            ignoreSigningInformation.set(options.ignoreSigningInformation)
```

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkPlugin.kt:
```diff
@@ -10,21 +13,68 @@ open class JLinkPlugin : Plugin<Project> {
+            compressionLevel.set(provider { options.compressionLevel.orNull })
+            endianness.set(provider { options.endianness.orNull })
+            ignoreSigningInformation.set(provider { options.ignoreSigningInformation.orNull })
+            excludeHeaderFiles.set(provider { options.excludeHeaderFiles.orNull })
```
```suggestion
            excludeHeaderFiles.set(options.excludeHeaderFiles)
```

Ect.....

SamCarlberg replied:
Done

JLLeitschuh commented on samples/groovy-dsl/build.gradle:
```diff
@@ -0,0 +1,24 @@
+
+jlink {
+    register("release image") {
+        applicationJar.set(provider { tasks.withType(Jar).getByName("jar").archivePath })
```
This doesn't convey the dependency information the way you want. You need to use the suggestion I provided above, (may require Gradle 5.0 to work).

SamCarlberg replied:
I'd prefer not to have to rely on high versions of Gradle, though it seems 5.0 is near to release

JLLeitschuh commented on samples/javafx-app/build.gradle.kts:
```diff
@@ -0,0 +1,66 @@
+jlink {
+    "release image" {
+        useMinimalImage()
+        applicationJar.set(provider { jarTask.archivePath })
```
Same issue here with not conveying the dependency information.

JLLeitschuh commented on samples/javafx-app/build.gradle.kts:
```diff
@@ -0,0 +1,66 @@
+    if (name == "jlinkGenerateReleaseImage") {
+        dependsOn("jar")
+    }
+}
```
If we set up this plugin correctly, this shouldn't be necessary.

JLLeitschuh committed:
Minor fix in README

Co-Authored-By: SamCarlberg <sam.carlberg@gmail.com>

SamCarlberg committed:
Use direct property bindings instead of wrapping in providers

JLLeitschuh commented on build.gradle.kts:
```diff
@@ -16,4 +19,11 @@ project(":samples") {
+    dependsOn(gradle.includedBuild("simple-jlink").task(":check"))
+    dependsOn(findProject(":samples:javafx-app")!!.tasks.getByName("jlinkGenerateReleaseImage"))
+    dependsOn(findProject(":samples:simple-jar")!!.tasks.getByName("jlinkGenerateReleaseImage"))
+    dependsOn(findProject(":samples:groovy-dsl")!!.tasks.getByName("jlinkGenerateReleaseImage"))
```
Make each sub-project's `check` task depend upon that project's `jlinkGenerateReleaseImage ` task, not the root `check` task.

SamCarlberg replied:
I'd prefer the samples to be normal projects that use the plugin.

JLLeitschuh replied:
These goals aren't contradictory?

SamCarlberg replied:
Well, the way I see it is that the root project just ties the plugin project and the sample projects together. So having the root project's `check` task make sure the samples play well with the plugin makes more sense to me than having each sample project's standard `check` task not only check the normal project things (tests, static analysis, etc) but also check the generated jlink distributions

JLLeitschuh commented on plugin/src/main/kotlin/org/gradleweaver/plugins/jlink/JLinkLauncherOptions.kt:
```diff
@@ -0,0 +1,43 @@
+     * The name of the generated launcher script. If not set, the launcher name will be the name of the
+     * corresponding project.
+     */
+    val launcherName = objectFactory.property(project.name)
```
Make the default lazily evaluate `project.name`.

SamCarlberg replied:
Does it matter? I thought the project name will always be defined since it's an immutable property

JLLeitschuh replied:
Oh, if it's immutable, then it's fine.

SamCarlberg committed:
Use java.home property to determine tools path

SamCarlberg committed:
Make root check task rely on all jlink tasks from the samples

Instead of hardcoding for the task names

SamCarlberg committed:
Minor cleanup

SamCarlberg committed:
Link project module into image, make plugin apply application plugin

For simplicity's sake, make JLinkTasks always depend on the project's jar task (for determining module info and linking/bundling). Support for custom jar files may be re-implemented in the future
Make the JavaFX sample modular
Add a simple modular project sample
Make task caching work (which unfortunately requires manually implementing Serializable)

SamCarlberg committed:
Flatten launcher options into JLinkOptions

SamCarlberg commented:
Merging this PR as-is. Future issues will be addressed in followup PRs. ATM the plugin is functional and usable, and I don't want to make this PR stay open for ages (it's already been a month!) while small improvements are made.
